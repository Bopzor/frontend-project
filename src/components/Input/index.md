`<Input />` allow users to enter text into a UI. They typically appear in forms and dialogs.

```js
<div style={{ width: 275 }}>
  <Input defaultValue="H&M" />
</div>
```

```js
<div style={{ width: 275 }}>
  <Input placeholder="Placeholder" />
</div>
```

```js
<div style={{ width: 275 }}>
  <Input disabled readOnly value="Disabled" />
</div>
```

You can specify `prefix` and `suffix` prop to create input adornments and indicate, for example, a search field.

```js
const Search = require('../Icons/Search').default;
const Close = require('../Icons/Close').default;
const IconButton = require('../IconButton').default;
<div style={{ width: 275 }}>
  <Input
    defaultValue="Plu"
    prefix={<Search color="#bfc3da" />}
    suffix={
      <IconButton backgroundColor="transparent" onClick={() => alert('You clicked me!')}>
        <Close color="#bfc3da" />
      </IconButton>
    }
  />
</div>;
```

We have a special `password` `<Input />` which places a suffix by default which allows it to trigger password visibility.

```js
  <div style={{ width: 275 }}>
    <Input defaultValue="password" type="password" />
  </div>
```

You can choose text color by using `variant` prop.

```js
<div style={{ display: 'flex', flexDirection: 'column', maxWidth: 275 }}>
  <Input variant="success" defaultValue="Success" />
  <br />
  <Input variant="warning" defaultValue="Warning" />
  <br />
  <Input variant="alert" defaultValue="Alert" />
</div>
```
