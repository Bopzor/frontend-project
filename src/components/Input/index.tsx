import * as React from 'react';

import { hot } from 'react-hot-loader';

import { css, cx } from 'emotion';
import memoizeOne from 'memoize-one';
import { Omit } from 'react-redux';

import { Consumer, Theme } from '../../theme';
import IconButton from '../IconButton';
import View from '../Icons/View';
import ViewOff from '../Icons/ViewOff';

/**
 * Props for the Input component
 */
export type InputProps = {
  /**
   * Text color
   */
  variant?: keyof Theme['colors'];
  /**
   * Element to appear as prefix adornment
   */
  prefix?: React.ReactNode;
  /**
   * Element to appear as suffix adornment
   */
  suffix?: React.ReactNode;
  /**
   * Ref to the underlying input component (just in case)
   */
  innerRef?: React.Ref<HTMLInputElement>;
} & Omit<React.HTMLProps<HTMLInputElement>, 'prefix'>;

export type InputState = {
  showPassword?: boolean;
};

/**
 * Allows to render a styled Input
 * Please refer to the styleguide for more info
 */
class Input extends React.Component<InputProps, InputState> {
  state: InputState = {};

  /**
   * Allows to toggle password visibility
   * To bind to password type default suffix
   */
  toggleShowPassword = () => {
    this.setState((state) => ({ showPassword: !state.showPassword }));
  };

  /**
   * Resolves the appropriate type of input (allows to trigger password visibility)
   */
  get type() {
    const { type } = this.props;
    const { showPassword } = this.state;
    if (!type || type !== 'password') {
      return type;
    }
    return showPassword ? 'text' : type;
  }

  /**
   * Get the supplied suffix or default suffix View/ViewOff for password type inputs
   */
  get suffix() {
    const { type, suffix } = this.props;
    const { showPassword } = this.state;
    if (!suffix && type === 'password') {
      return (
        <IconButton onClick={this.toggleShowPassword} backgroundColor="transparent" color="placeholder">
          {showPassword ? <ViewOff /> : <View />}
        </IconButton>
      );
    }
    return suffix;
  }

  render() {
    const {
      variant = 'text',
      suffix: _suffix,
      type: _type,
      prefix,
      className,
      innerRef,
      ...props
    } = this.props;
    const { suffix, type } = this;
    return (
      <Consumer>
        {(theme) => {
          const classNames = cx(wrapperStyles(theme), className);
          const inputClasses = cx(inputStyles(theme, variant), !!prefix && inputWithPrefixStyles);
          return (
            <div className={classNames}>
              {prefix && <span className={adornmentStyles}>{prefix}</span>}
              <input type={type} className={inputClasses} ref={innerRef} {...props} />
              {suffix && <span className={adornmentStyles}>{suffix}</span>}
            </div>
          );
        }}
      </Consumer>
    );
  }
}

export default hot(module)(Input);

/**
 * Styles for the div wrapper of input
 */
const wrapperStyles = memoizeOne(({ colors }: Theme) => {
  return css`
    display: flex;
    width: 100%;
    height: 44px;
    border-radius: 10px;
    background-color: ${colors.input};
    align-items: center;
    align-content: center;
    border: 2px solid transparent;
    :focus-within {
      border: 2px solid ${colors.focused};
    }
  `;
});
/**
 * Controls input appearence
 */
const inputStyles = memoizeOne(({ colors }: Theme, variant: InputProps['variant']) => {
  return css`
    flex: 1;
    background-color: ${colors.input};
    max-width: calc(100% - 75px);
    font-size: 15px;
    line-height: 1.33;
    color: ${colors[variant!]};
    border-radius: 10px;
    margin: 0 14px;
    border: unset;
    caret-color: ${colors.melrose};
    :focus {
      outline: none;
    }
    :disabled {
      color: ${colors.disabled};
    }
    ::placeholder {
      color: ${colors.placeholder};
    }
  `;
});
/**
 * Adornments (suffix & prefix) styles
 */
const adornmentStyles = css`
  margin: 0 14px;
`;
/**
 * Patch for input with has a prefix
 */
const inputWithPrefixStyles = css`
  margin: 0;
`;
