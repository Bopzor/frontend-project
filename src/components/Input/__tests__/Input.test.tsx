import * as React from 'react';

import { mount, render } from 'enzyme';
import { renderToJson } from 'enzyme-to-json';

import Input from '../';
import InputButton from '../../IconButton';

describe('Input', () => {
  describe('Should match snapshot for', () => {
    test('default', () => {
      const cheerio = render(<Input />);
      expect(renderToJson(cheerio)).toMatchSnapshot();
    });
    test('disabled', () => {
      const cheerio = render(<Input disabled />);
      expect(renderToJson(cheerio)).toMatchSnapshot();
    });
    test('with prefix', () => {
      const cheerio = render(<Input prefix="P" />);
      expect(renderToJson(cheerio)).toMatchSnapshot();
    });
    test('with suffix', () => {
      const cheerio = render(<Input />);
      expect(renderToJson(cheerio)).toMatchSnapshot();
    });
    test('default password type', () => {
      const cheerio = render(<Input type="password" />);
      expect(renderToJson(cheerio)).toMatchSnapshot();
    });
    test('password type with custom suffix', () => {
      const cheerio = render(<Input type="password" suffix="S" />);
      expect(renderToJson(cheerio)).toMatchSnapshot();
    });
    test('with variants', () => {
      const cheerio = render(
        <div>
          <Input variant="alert" />
          <Input variant="success" />
          <Input variant="primary" />
          <Input variant="text" />
        </div>
      );
      expect(renderToJson(cheerio)).toMatchSnapshot();
    });
  });
  test('Should toggle visibility for password', () => {
    const wrapper = mount(<Input type="password" />);
    expect(
      wrapper
        .find('input')
        .getDOMNode()
        .getAttribute('type')
    ).toBe('password');
    wrapper.find(InputButton).simulate('click');
    expect(
      wrapper
        .find('input')
        .getDOMNode()
        .getAttribute('type')
    ).toBe('text');
  });
});
