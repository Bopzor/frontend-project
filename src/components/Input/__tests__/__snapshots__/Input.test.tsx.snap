// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Should match snapshot for default 1`] = `
.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 44px;
  border-radius: 10px;
  background-color: #f1f2f8;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  border: 2px solid transparent;
}

.emotion-1:focus-within {
  border: 2px solid #9ee1ff;
}

.emotion-0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  background-color: #f1f2f8;
  max-width: calc(100% - 75px);
  font-size: 15px;
  line-height: 1.33;
  color: #2a225b;
  border-radius: 10px;
  margin: 0 14px;
  border: unset;
  caret-color: #908eff;
}

.emotion-0:focus {
  outline: none;
}

.emotion-0:disabled {
  color: #ced1e2;
}

.emotion-0::-webkit-input-placeholder {
  color: #9fa6c7;
}

.emotion-0::-moz-placeholder {
  color: #9fa6c7;
}

.emotion-0:-ms-input-placeholder {
  color: #9fa6c7;
}

.emotion-0::placeholder {
  color: #9fa6c7;
}

<div
  class="emotion-1"
>
  <input
    class="emotion-0"
  />
</div>
`;

exports[`Input Should match snapshot for default password type 1`] = `
.emotion-2 {
  margin: 0 14px;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 44px;
  border-radius: 10px;
  background-color: #f1f2f8;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  border: 2px solid transparent;
}

.emotion-3:focus-within {
  border: 2px solid #9ee1ff;
}

.emotion-0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  background-color: #f1f2f8;
  max-width: calc(100% - 75px);
  font-size: 15px;
  line-height: 1.33;
  color: #2a225b;
  border-radius: 10px;
  margin: 0 14px;
  border: unset;
  caret-color: #908eff;
}

.emotion-0:focus {
  outline: none;
}

.emotion-0:disabled {
  color: #ced1e2;
}

.emotion-0::-webkit-input-placeholder {
  color: #9fa6c7;
}

.emotion-0::-moz-placeholder {
  color: #9fa6c7;
}

.emotion-0:-ms-input-placeholder {
  color: #9fa6c7;
}

.emotion-0::placeholder {
  color: #9fa6c7;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 50%;
  padding: 0;
  border: none;
  width: 24px;
  height: 24px;
  color: #9fa6c7;
  cursor: pointer;
  pointer-events: all;
  background: transparent;
}

.emotion-1 > * {
  width: 20px;
  height: 20px;
}

<div
  class="emotion-3"
>
  <input
    class="emotion-0"
    type="password"
  />
  <span
    class="emotion-2"
  >
    <button
      class="emotion-1"
    >
      <svg
        fill="currentColor"
        height="24"
        viewBox="0 0 32 32"
        width="24"
      >
        <path
          d="M16.003 24.01c-7.338 0-10.983-5.836-11.89-7.539a.974.974 0 0 1 .003-.927l-.003.005c.907-1.703 4.552-7.539 11.89-7.539s10.983 5.836 11.89 7.539a.974.974 0 0 1-.002.927l.003-.005c-.907 1.703-4.552 7.539-11.89 7.539zm.007-13.5a5.5 5.5 0 1 0 5.5 5.5 5.5 5.5 0 0 0-5.5-5.5zm0 9.001a3.5 3.5 0 1 1 3.5-3.5 3.5 3.5 0 0 1-3.5 3.5z"
        />
      </svg>
    </button>
  </span>
</div>
`;

exports[`Input Should match snapshot for disabled 1`] = `
.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 44px;
  border-radius: 10px;
  background-color: #f1f2f8;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  border: 2px solid transparent;
}

.emotion-1:focus-within {
  border: 2px solid #9ee1ff;
}

.emotion-0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  background-color: #f1f2f8;
  max-width: calc(100% - 75px);
  font-size: 15px;
  line-height: 1.33;
  color: #2a225b;
  border-radius: 10px;
  margin: 0 14px;
  border: unset;
  caret-color: #908eff;
}

.emotion-0:focus {
  outline: none;
}

.emotion-0:disabled {
  color: #ced1e2;
}

.emotion-0::-webkit-input-placeholder {
  color: #9fa6c7;
}

.emotion-0::-moz-placeholder {
  color: #9fa6c7;
}

.emotion-0:-ms-input-placeholder {
  color: #9fa6c7;
}

.emotion-0::placeholder {
  color: #9fa6c7;
}

<div
  class="emotion-1"
>
  <input
    class="emotion-0"
    disabled=""
  />
</div>
`;

exports[`Input Should match snapshot for password type with custom suffix 1`] = `
.emotion-1 {
  margin: 0 14px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 44px;
  border-radius: 10px;
  background-color: #f1f2f8;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  border: 2px solid transparent;
}

.emotion-2:focus-within {
  border: 2px solid #9ee1ff;
}

.emotion-0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  background-color: #f1f2f8;
  max-width: calc(100% - 75px);
  font-size: 15px;
  line-height: 1.33;
  color: #2a225b;
  border-radius: 10px;
  margin: 0 14px;
  border: unset;
  caret-color: #908eff;
}

.emotion-0:focus {
  outline: none;
}

.emotion-0:disabled {
  color: #ced1e2;
}

.emotion-0::-webkit-input-placeholder {
  color: #9fa6c7;
}

.emotion-0::-moz-placeholder {
  color: #9fa6c7;
}

.emotion-0:-ms-input-placeholder {
  color: #9fa6c7;
}

.emotion-0::placeholder {
  color: #9fa6c7;
}

<div
  class="emotion-2"
>
  <input
    class="emotion-0"
    type="password"
  />
  <span
    class="emotion-1"
  >
    S
  </span>
</div>
`;

exports[`Input Should match snapshot for with prefix 1`] = `
.emotion-0 {
  margin: 0 14px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 44px;
  border-radius: 10px;
  background-color: #f1f2f8;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  border: 2px solid transparent;
}

.emotion-2:focus-within {
  border: 2px solid #9ee1ff;
}

.emotion-1 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  background-color: #f1f2f8;
  max-width: calc(100% - 75px);
  font-size: 15px;
  line-height: 1.33;
  color: #2a225b;
  border-radius: 10px;
  margin: 0 14px;
  border: unset;
  caret-color: #908eff;
  margin: 0;
}

.emotion-1:focus {
  outline: none;
}

.emotion-1:disabled {
  color: #ced1e2;
}

.emotion-1::-webkit-input-placeholder {
  color: #9fa6c7;
}

.emotion-1::-moz-placeholder {
  color: #9fa6c7;
}

.emotion-1:-ms-input-placeholder {
  color: #9fa6c7;
}

.emotion-1::placeholder {
  color: #9fa6c7;
}

<div
  class="emotion-2"
>
  <span
    class="emotion-0"
  >
    P
  </span>
  <input
    class="emotion-1"
  />
</div>
`;

exports[`Input Should match snapshot for with suffix 1`] = `
.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 44px;
  border-radius: 10px;
  background-color: #f1f2f8;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  border: 2px solid transparent;
}

.emotion-1:focus-within {
  border: 2px solid #9ee1ff;
}

.emotion-0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  background-color: #f1f2f8;
  max-width: calc(100% - 75px);
  font-size: 15px;
  line-height: 1.33;
  color: #2a225b;
  border-radius: 10px;
  margin: 0 14px;
  border: unset;
  caret-color: #908eff;
}

.emotion-0:focus {
  outline: none;
}

.emotion-0:disabled {
  color: #ced1e2;
}

.emotion-0::-webkit-input-placeholder {
  color: #9fa6c7;
}

.emotion-0::-moz-placeholder {
  color: #9fa6c7;
}

.emotion-0:-ms-input-placeholder {
  color: #9fa6c7;
}

.emotion-0::placeholder {
  color: #9fa6c7;
}

<div
  class="emotion-1"
>
  <input
    class="emotion-0"
  />
</div>
`;

exports[`Input Should match snapshot for with variants 1`] = `
.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 44px;
  border-radius: 10px;
  background-color: #f1f2f8;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  border: 2px solid transparent;
}

.emotion-1:focus-within {
  border: 2px solid #9ee1ff;
}

.emotion-6 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  background-color: #f1f2f8;
  max-width: calc(100% - 75px);
  font-size: 15px;
  line-height: 1.33;
  color: #2a225b;
  border-radius: 10px;
  margin: 0 14px;
  border: unset;
  caret-color: #908eff;
}

.emotion-6:focus {
  outline: none;
}

.emotion-6:disabled {
  color: #ced1e2;
}

.emotion-6::-webkit-input-placeholder {
  color: #9fa6c7;
}

.emotion-6::-moz-placeholder {
  color: #9fa6c7;
}

.emotion-6:-ms-input-placeholder {
  color: #9fa6c7;
}

.emotion-6::placeholder {
  color: #9fa6c7;
}

.emotion-0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  background-color: #f1f2f8;
  max-width: calc(100% - 75px);
  font-size: 15px;
  line-height: 1.33;
  color: #fd5573;
  border-radius: 10px;
  margin: 0 14px;
  border: unset;
  caret-color: #908eff;
}

.emotion-0:focus {
  outline: none;
}

.emotion-0:disabled {
  color: #ced1e2;
}

.emotion-0::-webkit-input-placeholder {
  color: #9fa6c7;
}

.emotion-0::-moz-placeholder {
  color: #9fa6c7;
}

.emotion-0:-ms-input-placeholder {
  color: #9fa6c7;
}

.emotion-0::placeholder {
  color: #9fa6c7;
}

.emotion-2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  background-color: #f1f2f8;
  max-width: calc(100% - 75px);
  font-size: 15px;
  line-height: 1.33;
  color: #24bdb9;
  border-radius: 10px;
  margin: 0 14px;
  border: unset;
  caret-color: #908eff;
}

.emotion-2:focus {
  outline: none;
}

.emotion-2:disabled {
  color: #ced1e2;
}

.emotion-2::-webkit-input-placeholder {
  color: #9fa6c7;
}

.emotion-2::-moz-placeholder {
  color: #9fa6c7;
}

.emotion-2:-ms-input-placeholder {
  color: #9fa6c7;
}

.emotion-2::placeholder {
  color: #9fa6c7;
}

.emotion-4 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  background-color: #f1f2f8;
  max-width: calc(100% - 75px);
  font-size: 15px;
  line-height: 1.33;
  color: #5856d6;
  border-radius: 10px;
  margin: 0 14px;
  border: unset;
  caret-color: #908eff;
}

.emotion-4:focus {
  outline: none;
}

.emotion-4:disabled {
  color: #ced1e2;
}

.emotion-4::-webkit-input-placeholder {
  color: #9fa6c7;
}

.emotion-4::-moz-placeholder {
  color: #9fa6c7;
}

.emotion-4:-ms-input-placeholder {
  color: #9fa6c7;
}

.emotion-4::placeholder {
  color: #9fa6c7;
}

<div>
  <div
    class="emotion-1"
  >
    <input
      class="emotion-0"
    />
  </div>
  <div
    class="emotion-1"
  >
    <input
      class="emotion-2"
    />
  </div>
  <div
    class="emotion-1"
  >
    <input
      class="emotion-4"
    />
  </div>
  <div
    class="emotion-1"
  >
    <input
      class="emotion-6"
    />
  </div>
</div>
`;
